src/writers/sink.creator.cpp

```diff
@@ -0,0 +1,409 @@
+#include "sink.creator.hh"
+#include "file.sink.hh"
+
+#include <latch>
+#include <queue>
+
+namespace zarr = acquire::sink::zarr;
+namespace common = zarr::common;
+
+zarr::SinkCreator::SinkCreator(std::shared_ptr<common::ThreadPool> thread_pool_)
+  : thread_pool_{ thread_pool_ }
+{
+}
+
+bool
+zarr::SinkCreator::create_chunk_sinks(
+  const std::string& base_uri,
+  const std::vector<Dimension>& dimensions,
+  std::vector<std::shared_ptr<Sink>>& chunk_sinks)
+{
+    return make_part_sinks_(
+      base_uri, dimensions, common::chunks_along_dimension, chunk_sinks);
+}
+
+bool
+zarr::SinkCreator::create_shard_sinks(
+  const std::string& base_uri,
+  const std::vector<Dimension>& dimensions,
+  std::vector<std::shared_ptr<Sink>>& shard_sinks)
+{
+    return make_part_sinks_(
+      base_uri, dimensions, common::shards_along_dimension, shard_sinks);
+}
+
+bool
+zarr::SinkCreator::create_v2_metadata_sinks(
+  const std::string& base_uri,
+  size_t n_arrays,
+  std::vector<std::shared_ptr<Sink>>& metadata_sinks)
+{
+    std::queue<std::string> dir_paths, file_paths;
+
+    file_paths.emplace(".metadata"); // base metadata
+    file_paths.emplace("0/.zattrs"); // external metadata
+    file_paths.emplace(".zattrs");   // group metadata
+
+    for (auto i = 0; i < n_arrays; ++i) {
+        dir_paths.push(std::to_string(i));
+        file_paths.push(std::to_string(i) + "/.zarray"); // array metadata
+    }
+
+    return make_metadata_sinks_(
+      base_uri, dir_paths, file_paths, metadata_sinks);
+}
+
+bool
+zarr::SinkCreator::create_v3_metadata_sinks(
```

Comment:
We might check that `base_uri` is nonempty, but if `n_arrays` is 0, then we simply don't create metadata files for each array.

Reply:
On the other hand, it might just be better to strip off a trailing `/` character, in case, for whatever reason, `base_uri` ends up being equal to `/`. Later use has `base_dir` constructed from `base_uri`, with paths constructed additively:

```
base_dir + "/" + dir_paths.front()
```

There's a potential "//" error in that case.